### PR TITLE
[Replicate] Add unit test with replicate param names

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102401

This attribute wasn't actually used in tests, add a test ensuring that
if replicate is used on top of FSDP, the replicated parameter names are as
expected.

TODO: there are a few ways to check if module is managed by composable API,
such as replicated param names for replicate, _get_module_state API,
_get_registry_api, etc. We should unify all composable APIs to check in a
unified way (filed an issue)

Differential Revision: [D46236377](https://our.internmc.facebook.com/intern/diff/D46236377/)

### CODE REVIEW

**Review by awgu — APPROVED** [review]

**Review by awgu — APPROVED** [review]
Stamp to unblock since this looks easy. Lint and unit test failures look real. I can take a deeper look by beginning of next week.

File: test/distributed/_composable/test_compose.py [hunk 1]
```diff
@@ -145,6 +147,33 @@ def test_checkpoint_fsdp_submodules_use_reentrant(self):
...
 
+    @skip_if_lt_x_gpu(2)
+    def test_fully_shard_replicate_correct_replicate_params(self):
+        model = CompositeParamModel(device=torch.device("cuda"))
+        # Shard Linears within UnitModule
+        fully_shard(model.u1, policy=ModuleWrapPolicy({nn.Linear}))
+        fully_shard(model.u2, policy=ModuleWrapPolicy({nn.Linear}))
+        # replicate the rest
+        replicate(model)
+        # Run fwd + bwd to initialize DDP
```
Comment (awgu) [inline test/distributed/_composable/test_compose.py:158]: Curious: what does it mean to initialize DDP? Do you have to run backward?

File: test/distributed/_composable/test_compose.py [hunk 2]
```diff
@@ -145,6 +147,33 @@ def test_checkpoint_fsdp_submodules_use_reentrant(self):
...
+        # immediate parameters of model and parameters of model's non-UnitModule
+        # submodules are replicated
+        param_names = replicate.state(model)._replicate_param_names
+        replicated_modules = [
+            (name, mod)
+            for (name, mod) in model.named_children()
+            if mod not in [model.u1, model.u2]
+        ]
+        replicated_param_names = [
+            f"{module_name}.{n}"
```
Comment (awgu) [inline test/distributed/_composable/test_compose.py:171]: So these are _not_ all prefixed from the same root module?

File: torch/distributed/_composable/replicate.py [hunk 3]
```diff
@@ -8,6 +8,8 @@
 
 from .contract import _get_registry, contract
 
+_ROOT_MODULE_PREFIX = ""
```
Comment (awgu) [inline torch/distributed/_composable/replicate.py:11]: Could you explain why we define this? Is this meant to be changed in the future? Will it ever be something other than the empty string?

Comment (rohan-varma) [reply torch/distributed/_composable/replicate.py:11]: it is mostly so I don't hardcode `""` a few different times in the recursive name generation logic.

File: test/distributed/_composable/test_compose.py (hunk 1 above)
Comment (rohan-varma) [reply test/distributed/_composable/test_compose.py:158]: In replicate, DDP is only initialized in the first forward. You don't have to run backward, will clarify that in the comment

File: test/distributed/_composable/test_compose.py (hunk 2 above)
Comment (rohan-varma) [reply test/distributed/_composable/test_compose.py:171]: For something like
```
class MyModel:
    def __init__(self):
        self.lin1 = ...
        self.lin2 = ....
        self.p = nn.Parameter(...)
```

and if we replicate the whole thing, the names we expect are like `[lin1.weight, lin2.weight, p]`. They are all prefixed from the same root, but the root is `MyModel` and the root prefix is just `""`.

**Review by awgu — APPROVED** [review]

**Pull request closed without merging**